shell/browser/api/electron_api_desktop_capturer.cc

```diff
@@ -181,6 +181,42 @@ DesktopCapturer::DesktopCapturer(v8::Isolate* isolate) {}
 
 DesktopCapturer::~DesktopCapturer() = default;
 
+DesktopCapturer::DesktopListListener::DesktopListListener(
+    OnceCallback update_callback,
+    OnceCallback failure_callback,
+    bool skip_thumbnails) {
+  update_callback_ = std::move(update_callback);
+  failure_callback_ = std::move(failure_callback);
+  have_thumbnail_ = skip_thumbnails;
+}
+
+DesktopCapturer::DesktopListListener::~DesktopListListener() = default;
+
+void DesktopCapturer::DesktopListListener::OnDelegatedSourceListSelection() {
+  if (have_thumbnail_) {
+    std::move(this->update_callback_).Run();
+  } else {
+    have_selection_ = true;
+  }
+}
+
+void DesktopCapturer::DesktopListListener::OnSourceThumbnailChanged(int index) {
+  if (have_selection_) {
+    // This is called every time a thumbnail is refreshed. Reset variable to
+    // ensure that the callback is not run again.
+    have_selection_ = false;
+
+    // PipeWire returns a single source, so index is not relevant.
+    std::move(this->update_callback_).Run();
```

Comment:
This new logic applies cross-platform, though - is there a situation where we'd need this index?

Reply:
Not unless Chromium makes significant changes/refactors. PipeWire is the only delegated capturer right now. As a result, this code path will not be executed on other platforms.

Secondly, [according to Chromium documentation](https://source.chromium.org/chromium/chromium/src/+/refs/heads/main:chrome/browser/media/webrtc/desktop_media_list.h;l=122-128;drc=89758767fb0fae96a59967eb47e7921ce49fafcb), delegated capturers manage their own UI for source selection and will return a single source. This logic should hold even if there's a new delegated capturer in the future.